docs/rules/performance/defer-assignment.md

```diff
@@ -0,0 +1,79 @@
+# defer-assignment
+
+**Summary**: Assignment can be deferred
+
+**Category**: Performance
+
+**Avoid**
+```rego
+package policy
+
+import rego.v1
+
+allow if {
+    resp := http.send({"method": "GET", "url": "http://example.com"})
+
+    # this check does not depend on the response above
+    # and thus the resp := ... assignment can be deferred to
+    # after the check
+    input.user.name in allowed_users
+
+    resp.status_code == 200
+
+    # more done with response here
+}
+```
+
+**Prefer**
+```rego
+package policy
+
+import rego.v1
+
+allow if {
+    input.user.name in allowed_users
+
+    # the next expression *does* depend on `resp`
+    resp := http.send({"method": "GET", "url": "http://example.com"})
+
+    resp.status_code == 200
+
+    # more done with response here
+}
+```
+
+## Rationale
+
+Assignments are normally cheap, but certainly not always. If the right-hand side of an assignment is expensive,
+deferring the assignment to where it's needed can save a considerable amount of time. Even for less expensive
+assignments, code tends to be more readable when assignments are placed close to where they're used.
```

Comment:
> code tends to be more readable when assignments are placed close to where they're used.

I think this is the number one reason, cheap/expensive is the secondary reason for this rule since it does not always apply / is more complicated.

Something like this might work:

> Placing assignments close to where they’re used improves code readability. Following this rule is also particularly important when the right-hand side of an assignment is expensive, such as making an HTTP call, as deferring the assignment can also save significant time.

Reply:
This rule is in the performance category, which is why it's emphasized here. I think there are good examples of both really, and there are of course examples of both where location doesn't matter much.